### ISSUE TITLE
PBS/torque executor is using a PBS/pro argument for job array submission

### ISSUE THREAD

**Issue by madscort** [nextflow-io/nextflow#5218]
### Expected behavior and actual behavior

Using the new job array submission feature in combination with the PBS/torque executor does not work.
The PbsExecutor.groovy script adds the "-J" flag for array request to the qsub command.
This option would be correct for PBS/pro, but doesn't work for torque (v. 6.1.3). I think the correct command would be "-t" instead.

### Program output 

```
Failed to submit process to grid scheduler for execution


Command executed:

 qsub -N nf-CALLING_HAPL .command.run

Command exit status:
 2

Command output:
 [The -J option can only be used in conjunction with -P]

 usage: qsub [-a date_time] [-A account_string] [-b secs]
   [-c [ none | { enabled | periodic | shutdown |
   depth=<int> | dir=<path> | interval=<minutes>}... ]
   [-C directive_prefix] [-d path] [-D path]
   [-e path] [-h] [-I] [-j oe|eo|n] [-k {oe}] [-K <kill delay seconds>]
   [-l resource_list] [-m n|{abe}] [-M user_list] [-N jobname] [-o path]
   [-p priority] [-P proxy_user [-J <jobid]] [-q queue] [-r y|n]
   [-S path] [-t number_to_submit] [-T type] [-u user_list] [-w] path
   [-W additional_attributes] [-v variable_list] [-V ] [-x] [-X] [-z] [script]
```

### Environment 

* Nextflow version: 24.04.2
* Java version: 1.8.0
* Operating system: Linux
* Bash version: 4.2.46

**Comment by pditommaso** [comment]
Any chance you can validate the correct command and provide it to us?

**Comment by bentsherman** [comment]
I think I found the PBS/Torque docs, looks like everything we need is here: https://docs.adaptivecomputing.com/torque/4-0-2/help.htm#topics/commands/qsub.htm

I will draft a PR

**Comment by madscort** [comment]
Thanks a lot for fixing this! It does indeed work as expected, however I think the syntax for the pbs executor is still partly incorrect. Now jobs are submitted as they should, but I think something goes wrong in the monitoring. Very short jobs succeed, but longer ones fail with the:

`Process XX terminated for an unknown reason -- Likely it has been terminated by the external system
`

Even though they are in fact running on the HPC. This is not something I have experienced before and turning arrays off solves the problem. We only have one queue, so jobs being moved between queues is not the problem.

The way the executor monitors is by querying with `qstat -q -1` right? How does it keep track of individual jobs? Running only `qstat -q -1`, as an example gives you:

`Job Id: 49652838[].i-07-moab1.eth.clb
    job_state = C
`

So only a single status for the entire array, if you want the status for all individual jobs in each array, you need to add `-t` as well here. `qstat -q -1 -t` will in this case give you:

`Job Id: 49652838[0].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[1].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[2].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[3].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[4].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[5].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[6].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[7].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[8].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[9].i-07-moab1.eth.clb
    job_state = C
Job Id: 49652838[10].i-07-moab1.eth.clb
    job_state = C
`

Could this be the problem?

**Comment by bentsherman** [comment]
Nextflow relies on being able to query the status of individual child jobs in an array, so that it can query them like any other job. Most grid executors support this, here is what we have for PBS and PBSPro:

https://github.com/nextflow-io/nextflow/blob/6e866ae81ff3bf8a9729e9dbaa9dd89afcb81a4b/modules/nextflow/src/main/groovy/nextflow/executor/PbsExecutor.groovy#L199-L203

So we assume that `qstat -f -1 <jobid>[<index>]` should give us the status of the one child job. If that's not how it works then it could be a problem

**Comment by asp8200** [comment]
We had what seems to be the same problem on our HPC with Nextflow v25.10.4 and PBS/Torque v7.1.0. I seem to have solved it locally by changing `qstat -f -1` to `qstat -f -1 -t` [here](https://github.com/nextflow-io/nextflow/blob/6e866ae81ff3bf8a9729e9dbaa9dd89afcb81a4b/modules/nextflow/src/main/groovy/nextflow/executor/PbsExecutor.groovy#L141).

Do you want me to make a PR with that little change?

**Comment by nicorap** [comment]
Great @asp8200 , that helps a lot indeed.

**Comment by asp8200** [comment]
> We had what seems to be the same problem on our HPC with Nextflow v25.10.4 and PBS/Torque v7.1.0. I seem to have solved it locally by changing `qstat -f -1` to `qstat -f -1 -t` [here](https://github.com/nextflow-io/nextflow/blob/6e866ae81ff3bf8a9729e9dbaa9dd89afcb81a4b/modules/nextflow/src/main/groovy/nextflow/executor/PbsExecutor.groovy#L141).
> 

That fix paved the way for hitting the next issue:

"TaskArrayRun.getContainerConfig() was mutating the shared envWhitelist list on the SingularityConfig object directly while another thread was iterating over it. Fixed by creating a defensive copy of the list and returning a new SingularityConfig instance instead of mutating the shared one."

That is I made the following change:
```
% git diff modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayRun.groovy
diff --git a/modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayRun.groovy b/modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayRun.groovy
index 9640c9c71..e32c7d3e8 100644
--- a/modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayRun.groovy
+++ b/modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayRun.groovy
@@ -19,6 +19,7 @@ package nextflow.processor
 import groovy.transform.CompileStatic
 import groovy.util.logging.Slf4j
 import nextflow.container.ContainerConfig
+import nextflow.container.SingularityConfig
 import nextflow.executor.TaskArrayExecutor

 /**
@@ -41,9 +42,27 @@ class TaskArrayRun extends TaskRun {
     ContainerConfig getContainerConfig() {
         final config = super.getContainerConfig()
         final envWhitelist = config.getEnvWhitelist()
-        if( envWhitelist != null ) {
-            final executor = (TaskArrayExecutor)processor.getExecutor()
-            envWhitelist.add(executor.getArrayIndexName())
+        if( envWhitelist == null )
+            return config
+        final executor = (TaskArrayExecutor)processor.getExecutor()
+        final newList = new ArrayList<String>(envWhitelist)
+        newList.add(executor.getArrayIndexName())
+        if( config instanceof SingularityConfig ) {
+            final opts = [
+                autoMounts: config.autoMounts,
+                cacheDir: config.cacheDir,
+                enabled: config.enabled,
+                engineOptions: config.engineOptions,
+                envWhitelist: newList,
+                libraryDir: config.libraryDir,
+                noHttps: config.noHttps,
+                ociAutoPull: config.ociAutoPull,
+                ociMode: config.ociMode,
+                pullTimeout: config.pullTimeout,
+                registry: config.registry,
+                runOptions: config.runOptions
+            ]
+            return new SingularityConfig(opts)
         }
         return config
```

That was my fix for NF v25.10.4 (using Singularity), but it might also be an issue in more recent versions of NF.
